### DISCUSSION TITLE
Adapting the ARI format for ITRSs

### DISCUSSION THREAD

**Discussion by ffrohn** [TermCOMP/starexec-master#89] [General]
ITRSs are de facto a special case of LCTRSs, for which [ARI has a dedicated format](http://project-coco.uibk.ac.at/ARI/lctrs.php). To avoid a proliferation of different formats, it could be good to use this LCTRS format also for ITRSs. Termination analysis of LCTRSs is supported, e.g., by the tool [Ctrl](http://cl-informatik.uibk.ac.at/software/ctrl/), and there is active work on the tools [Cora](https://github.com/hezzel/cora/) and [Crisys2](https://git.trs.css.i.nagoya-u.ac.jp/nishida/crisys2/).

_Originally posted by @cfuhs in https://github.com/TermCOMP/starexec-master/discussions/85#discussioncomment-8144994_

**Comment by ffrohn** [comment]
True, but LCTRSs are very expressive. They allow...

1. other theories
2. Boolean variables
3. disjunction / negation

Dealing with 1. is easy, we just fix the theory (`QF_LIA`? `QF_NIA`?). Regarding 2. and 3., I'm not that familiar with our ITRSs. Are these things allowed in ITRSs as well? Are there other features of LCTRSs that might be problematic? What kind of variable conditions do both formalisms impose (i.e., where do they allow extra variables that do not occur on the left-hand side)?

Other "nasty" feature of SMT-Lib (and hence LCTRSs) are `let`-expressions, `if-then-else`, somewhat unexpected functions like `divides c` where `c > 0`, ... So if we go for LCTRSs, we might want to restrict ourselves to a subset of SMT-Lib...?

**Reply by AkihisaYamada** [reply]
"divides" is really nasty, it's used in form `((_ divides 2) x)`. Certainly we should restrict the format very well.
BTW: "theory" here should be `Ints`: http://smtlib.cs.uiowa.edu/theories.shtml.

**Reply by stefandollase** [reply]
I am currently aware of the following differences and interesting similarities.

* LCTRS have the same variable condition as TRSs, i.e., each variable that occurs on the LHS also has to occur on the RHS. ITRSs don't have this condition. Instead they impose a condition on the substitution.
* LCTRS rules don't have a cost, but ITRS rules do. If I didn't miss anything, then this is because there is currently no complexity analysis for LCTRS.
* LCTRS allow other theories. But for ITRSs, we could just fix the theory.
* LCTRS allow boolean variables, but ITRSs don't allow them.
* LCTRS allow disjunction / negation, just like ITRSs do.

To me, the biggest obstacle seems to be the different variable conditions and the missing cost annotations.

I will look into this more closely when we actually want to define the new ITRS format.

**Reply by ffrohn** [reply]
Which definition of ITRSs are you referring to? This one [1] seems to allow, e.g., Boolean variables, as far as I can see.

[1] https://doi.org/10.1007/978-3-642-02348-4_3

**Reply by stefandollase** [reply]
Indeed they are allowed in that paper. They didn't seem to be very useful to me, so I didn't consider them up to now. I will look into it again, when we define the new ITRS format.

**Reply by AkihisaYamada** [reply]
Assuming Boolean variables are not needed, a straightforward extension of ITS will be nice. Meaning just change
```
rule ::= ( rule term term guard? )
```

**Reply by stefandollase** [reply]
The following grammar seems to fit what I want to do.

```
         ITRS ::= ( format ITRS ) theory+ sort-decl* fun-decl* rule*
       theory ::= ( theory Ints )
    sort-decl ::= ( sort sort )
     fun-decl ::= ( fun identifier type )
         type ::= sort | ( -> sort+ sort )
         sort ::= identifier | Int
         rule ::= ( rule lhs rhs (:cost expression)? (:guard (and atom+))? )
          lhs ::= number | identifier | theory-symbol | ( theory-symbol lhs+ )
          rhs ::=          expression | theory-symbol | ( theory-symbol rhs+ )
theory-symbol ::= identifier
```

1. `expression` and `atom` are the same as for ITSs, see #90.
1. `expression` and `atom` cannot contain theory-symbols.
1. The left-hand side of a rule can neither be a variable nor a number.
1. For non-int-variables, all variables that occur on the right-hand side also have to occur on the left-hand side.
1. For int-variables, there is no such restriction.
   This allows for rules such as `f(x) -> y :|: y<x`.
1. Substitutions can neither contain int-operations nor theory-symbols or variables with return type int, i.e., if `x` is substituted by `t`, then all int-typed subterms of `t` are actual integers.
1. All variables in the cost or guard are int-typed.

**Reply by ffrohn** [reply]
Some comments:

Points 3. and 4. are in contrast to what's discussed [here](https://github.com/orgs/TermCOMP/discussions/92).

Regarding 7. and the definition of `rule`, I don't think we should restrict the format to conjunctions, forbid Booleans, etc. For [ITSs](https://github.com/orgs/TermCOMP/discussions/90), we'll presumably allow whatever the imported theory defines.

Regarding costs, @juergengiesl pointed out that allowing costs allows for modelling relative rewriting, such that dependency pairs no longer work. So for the termination category, we'll want to / have to forbid costs, most likely.

**Reply by jwaldmann** [reply]
I don't have all the details available, but if I am reading this right (I hope I'm not)

> .. allowing .. such that dependency pairs no longer work. So ... we'll want to / have to forbid ...

No. If one specific proof method does not work for a specific class of problems, then we don't declare these problems illegal, but instead we want to improve the method, or find another one, that does work.

Or do you mean that the DP method is need to define the semantics of problems?

**Reply by ffrohn** [reply]
> Or do you mean that the DP method is need to define the semantics of problems?

No. But so far, we do not have relative rules in the category `ITRS Termination`, and introducing them changes the problem substantially (I think all tools that ever participated -- `AProVE` and `Ctrl`? -- use DPs). So if we want relative rules, we should probably have a dedicated category (analogously to `TRS Standard` and `TRS Relative`).

**Reply by juergengiesl** [reply]
What I meant is that for TRSs, relative rules introduce subtle problems that require completely new proof techniques. The main problems are duplicating relative rules (so this is something that does not occur for SRSs and it also does not occur for ITSs where variables are instantiated by normal forms). The classical example is a TRS with a rule a -> b with cost 1, and a rule f(x) -> c(f(x),x) with cost 0. This TRS is not relatively terminating because of the sequence f(a) -> c(f(a),a) -> c(f(a),b) -> c(c(f(a),a),b) -> c(c(f(a),b),b) -> ...

There are more examples that show that proving termination of relative rules requires fundamentally different techniques than proving termination of ordinary rules. Therefore, it makes sense not add costs for the standard TRS category. And for the same reason it might also make sense not to add costs for ITRSs, to avoid making this even harder than it already is.

**Comment by ffrohn** [comment]
@stefandollase, are you still interested in this? Otherwise, I'd close this discussion for now.

**Reply by stefandollase** [reply]
You can close it for now. We can just reopen, when it becomes interesting again. Thanks for asking.
